### DISCUSSION TITLE
Global state management?

### DISCUSSION THREAD

**Discussion by yahia-berashish** [vanjs-org/van#152] [Q&A] [answered]
Hello again,
I'm currently toying with a utility project for VanJS, just experimenting at this point, and I have some questions. 

Is there currently a way to implement a global state not bound to a certain component or a single file?
Is it possible to consume state without passing it with props?
I will take anything, but I'm interested in the possibility of a Provider/Consumer approach, like the React Context API. 
This can potentially solve a lot of issues (e.g. prop drilling) and will make VanJS state management more advanced. 

I think it can be a bit complex with VanJS specifically because of its direct DOM manipulation approach, which makes updating the DOM at a global scale challenging - at least according to my time working with it/reading the docs -, but a possible solution is to use a custom prop attached to functional components which can be used to access and modify a state globally:
``` ts

export type Theme = "dark" | "light";

// `provide` function uses a key to inject the passed state to the component passed to the resulting function
const theme = van.state<Theme>("light");
const provideTheme: ProvideFunc<Theme> = provider("theme", theme);

export interface StatefulComponent<TProps = any, TStates extends Record<string, State> = Record<string, State>> {
  (props: TProps): ChildDom;
  state: TState
}

export const ThemeButton: StatefulComponent<any, { theme: State<Theme> }> = () => {
  // retrieve the `state` property of the caller function using a key
  // if the state isn't found, the function will look for it in the nearest caller function
  // the function can also inherit the callers state type
  const theme = getState("theme");

  // render something based on the current theme, toggle the current theme with an event
  return button({onclick: () => theme.val = theme.val === "dark" ? "light" : "dark",
   style: () => `background: ${theme.val === "light" ? "white" : "black"}`}, "Toggle Theme");
}

// inject state to component
provideTheme(ThemeButton)

// use the stateful component like any regular functional component
van.add(document.body, ThemeButton())

```

This is just some sketching code, I would like to get some opinions on the problem and how it could be solved.

**Comment by Tao-VanJS** [comment]
To be honest I'm not quite familiar with the React Context API. My understanding of context is a collection of states at the global scope with a set of ergonomic APIs to get and set their values. I think VanJS (which is a reactive UI framework) is largely orthogonal to global state management. I feel that it's quite easy to provide some simple implementation for a straight-forward approach (e.g.: via a global variable or something in the `window` object). It's also possible to have something more sophisticated if you want more features about context management. It should also be okay to integrate with some 3rd party library for it.

For the code you have shown, I think it's a reasonable approach. I've seen no problems with it.

**Comment by Hunter-Gu** [comment]
I think `van.state()` is the easiest way to do global state management. API like React Context is another way to do that, but it is a little different from global state management. React Context lets the parent component make some information available to any component in the tree below it, that really make sense when use create a group component (e.g. from group)

**Reply by yahia-berashish** [reply]
Well, I apologize for the confusion, but what I meant by **global state** is a state that is _not_ local i.e. not limited to a single component, but not necessarily available for the whole application.

**Reply by Hunter-Gu** [reply]
Got you. That means we want to have Context API in VanJS. 

I'm willing to help make it happen. Do you have any ideas about how to implement this in  VanJS?  It seems we need to maintain a tree structure in VanJS internally?

**Reply by yahia-berashish** [reply]
Appreciate your help!
The issue with implementing something like React Context API in VanJS is that React uses a virtual DOM: The DOM elements aren't manipulated directly, but a JavaScript object is used a virtual DOM to make the manipulations more performant, before a diffing algorithm is used to transfer the changes to the real DOM efficiently. 
The virtual DOM also allows React to pass states freely between different parts of the DOM tree because it is just a simple JavaScript Object, giving it immense flexibility. 

VanJS doesn't have a virtual DOM, making something like that a bit harder. 
What I think could be a potential solution is using a custom property in the functional component like I mentioned in the first message, maybe creating a provider component that will take a copy of it's children then add the state to that custom property (without modifying the original component), then providing a function that can be called inside a functional component to retrieve the state property of the caller function (the function which the state retrieval function was called inside) exposing it in a simple way. 

If you plan to work on the issue, maybe you can start from the sketch code I provided in the first message - just remember to open an issue.

**Reply by Hunter-Gu** [reply]
Thank you for your answer.

Actually I tried a way before, and the code is really simple

```ts
function createProvider<T>(defaultValue: T) {
  const data: T[] = [defaultValue];
  return {
    getProvider(): T {
      return data.slice(-1)[0];
    },
    Provider(val: T, children: any) {
      if (arguments.length === 1) {
        children = val;
        return children();
      }

      data.push(val);
      const res = children();
      data.pop();
      return res;
    },
  };
}

const { getProvider, Provider } = createProvider({ color: 'red' });
```

Try it on sandbox: https://codesandbox.io/p/sandbox/young-tdd-7xhky8?file=%2Fsrc%2Fmain.ts%3A1%2C1

The biggest problem is the component which depend on context value can't render in conditional, the component must be render together with Provider. And the component must get provided value at rendering.

Anyway, I will try to figure it out.

**Reply by yahia-berashish** [reply]
Looks good, is it reactive though (can the data be mutated)?
And passing children to the provider as an uncalled function doesn't work when you need to pass props to the component. 
Maybe you can use the `.caller` method of a functional component to get the provider that the component was inside, and using VanX to create a reactive store might be a good approach. 
I'm currently working on a prototype that uses a slightly different method, will share soon, but this is how the final API should look like: 
``` ts
const app = document.getElementById("app");
const { button } = van.tags;

type Theme = "dark" | "light";

interface ThemeState {
  theme: Theme
};

// pass type and default value to store constructor
const themeProvider = store<ThemeState>({ theme: "light" });

// declare with `StatefulComponent` type and pass state and props type
const ToggleTheme: StatefulComponent<ThemeState, { content: string }> = ({ content }) => {
  // a vanX.reactive() object, cannot use `const { theme } = ...`
  // the state function will look for the state in the nearest caller function, 
  // so the state has to be provided for a parent component only instead of passing it to every individual component
  const state = themeProvider.state();

  return button({ onclick: () => {state.theme = state.theme === "light" ? "dark" : "light"}, class: () => `${state.theme === "light" ? "light-btn" : "dark-btn"}`}, content);
};

// pass value (can be set to null to use default value) and components
themeProvider({ theme: "dark" }, ToggleTheme);

van.add(app, ToggleTheme({ content: "Toggle Theme" }));
```

**Reply by Hunter-Gu** [reply]
Yes, benefit from VanJS reactive state, the context data can be mutated. But user has to consume context in State-derived, maybe this is not intuitive enough.

**Reply by yahia-berashish** [reply]
I'm not sure what you mean by state-derived, if you mean the state has to be derived in order to be used in props, that's because it is reactive: 
``` ts
// reactive state has to be used inside derived function for attribute to be reactive
...button({ class: () => `${state.val === "red" ? "bg-red-500" : "bg-blue-500"}` }, "Reactive Button")...
```

This is kind of the cost of having reactive props, if the developer wants to use static data that cannot be mutated, then a regular JSON file can be more simple to implement.

**Reply by Hunter-Gu** [reply]
Hello again, I tried in another way. The key steps:

- create a class name and unique id for context provider, and store provide value
- find the ancestor with class name by `element.closest()` when use context 
- get the unique id from dom element id, then get the provide value from store

Drawback: it will always get the default context when component render, because the framework hasn't bind the reactive state and dom, and it's a little tricky to find the ancestor context provider.

Try it on sandbox: https://codesandbox.io/p/sandbox/vanjs-context-provider-poc-qsgdr8?file=%2Fsrc%2Fmain.ts%3A5%2C1

**Reply by yahia-berashish** [reply]
An interesting idea, but does `element.closest` get the nearest matching element or the nearest parent? Since inheriting state from a child element can be problematic. 
And I don't think I see where is the example you mentioned in the CodeSandbox link you provided, you might have shared the wrong project.

**Reply by Hunter-Gu** [reply]
I'm sorry, I copied the error link, updated now.

**Reply by yahia-berashish** [reply]
I have a few recommendations: 
- It may be better to allow the developer to pass the container as a prop to the provider instead of hard setting it to a div. 
- I noticed the returned value from `getProvider` is an object, it is preferred to use a VanX reactive object over a regular van state. 
- I don't seem to understand the purpose of creating different providers for the color and the setting function, can they be grouped in a single provider?

I think the code you wrote is quite impressive, maybe you can try to figure it out with @Tao-VanJS

**Reply by Hunter-Gu** [reply]
Thank you for your recommendations. It is still POC version now. I will try to design a friendly API once the plan accepted.

In the example, I create another providers for setting function is because we need to change the whole value of van state object. VanX reactive object can be better, I will replace the van state with VanX reactive object in the example.

**Reply by yahia-berashish** [reply]
Great 👍🏼
Removed the answered tag just to keep the discussion going. 
Would you open an issue please (just to keep the workflow clean) @Hunter-Gu

**Reply by Hunter-Gu** [reply]
I created an issue [161](https://github.com/vanjs-org/van/issues/161).

**Reply by yahia-berashish** [reply, marked as answer]
Good
So I say we close this Q&A
See you later 👋🏼

**Reply by Tao-VanJS** [reply]
Hi @yahia-berashish and @Hunter-Gu,

I appreciate your discussion in this thread. I don't think you need to close this thread. From the perspective of discoverability, keeping the thread here might be helpful for other people to find this interesting discussion :-)

**Reply by iuroc** [reply]
Sharing a bit about how I currently manage state using the following approach:

```typescript
export class SG<T> {
    public constructor(private init: T) { }

    public get<Key extends keyof T>(key: Key): T[Key] {
        return this.init[key]
    }
    public obj<Key extends keyof T>(key: Key) {
        return new SG(this.init[key])
    }
    public set<Key extends keyof T>(key: Key, value: T[Key]) {
        this.init[key] = value
    }
}

const sg = new SG({
    a: van.state(333),
    b: new SG({
        c: van.state(666)
    })
});

console.log(sg.get<number>('a'))  // { val: 333 }
console.log(sg.obj('b').get<number>('c'))  // { val: 666 }

const Component = () => {
    return div(
        sg.get('a'),
        button({
            onclick() {
                sg.get('a').val = 111;
            },
        }, 'Change a val')
    );
}
```

My project structure looks like this:

```
main.ts                  Entry file
view.ts                  Common components
route/                   Starting point for routes
    home/               Home route
        index.ts        Route component
        state.ts        Export SG (StateGroup)
        view.ts         Subcomponent
    about/
        index.ts        Route component
        state.ts        Export SG (StateGroup)
        view.ts         Subcomponent
```

**Reply by iuroc** [reply]
You can encapsulate it into a utility class.

```ts
/**
 * State Group
 */
export class SG<T> {
    public constructor(private init: T) { }

    public get<Key extends keyof T>(key: Key): T[Key] {
        return this.init[key]
    }
    public obj<Key extends keyof T>(key: Key) {
        return new SG(this.init[key])
    }
    public set<Key extends keyof T>(key: Key, value: T[Key]) {
        this.init[key] = value
    }
}
```

**Reply by yahia-berashish** [reply]
Hello @iuroc 
Thanks for your contribution, but using objects for states has been natively implemented with VanX (`vanjs-ext` to download via package manager). 
The current issue is concerned with handling global state in a growing web application. 
Currently the solutions we are focusing on are: 
- A React Context-style API
- A Svelte Stores/Observer-style API (which isn't really necessary because VanJS natively handles state to reactivity update the DOM, but it can be useful in larger apps)
- A Redux/Mobx-style API

Again, thanks for your contribution 👋🏼

**Reply by iuroc** [reply]
@yahia-berashish Thank you very much for your understanding. However, I am currently seeking a solution that does not involve VanX because I enjoy tinkering on my own. Thanks again; VanX is indeed a better option. 😉
